### PR TITLE
gh-102114: Make dis print more concise tracebacks for syntax errors in str inputs

### LINKED ISSUE #102114: dis._try_compile: traceback printed on error in source argument is too wordy
`_try_compile` function from `dis` module firstly tries to compile given source string with `'eval'` mode and, if exception is occured, catches it and tries again with `'exec'` mode. These actions lead to a long chained traceback if given string contains syntax/indentation error, because this string gets refused by `compile` in both `'eval'` and `'exec'` modes.
All functions/constructors from `dis` module that use `_try_compile` and accept source string as an argument (`dis`, `get_instructions`, `code_info`, `show_code`, `Bytecode`) are showing this behavior:
```python
>>> dis.dis(')')
Traceback (most recent call last):
  File "/home/.../cpython/Lib/dis.py", line 67, in _try_compile
    c = compile(source, name, 'eval')
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<dis>", line 1
    )
    ^
SyntaxError: unmatched ')'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/.../cpython/Lib/dis.py", line 112, in dis
    _disassemble_str(x, file=file, depth=depth, show_caches=show_caches, adaptive=adaptive)
  File "/home/.../cpython/Lib/dis.py", line 593, in _disassemble_str
    _disassemble_recursive(_try_compile(source, '<dis>'), **kwargs)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.../cpython/Lib/dis.py", line 69, in _try_compile
    c = compile(source, name, 'exec')
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<dis>", line 1
    )
    ^
SyntaxError: unmatched ')'
```
Python versions affected: 3.10.8, 3.11.0, current main.

### Linked PRs
* gh-102115

### PR DESCRIPTION
**Pull request by chgnrdv**
Fixes #102114

* Issue: gh-102114

### CODE REVIEW

**Review by iritkatriel — CHANGES_REQUESTED** [review]
Thanks for the PR, your change makes sense. However, we would need a unit test covering this behaviour to prevent future regressions.

File: Lib/test/test_dis.py [hunk 1]
```diff
@@ -1065,6 +1065,13 @@ def check(expected, **kwargs):
         check(dis_nested_2, depth=None)
         check(dis_nested_2)
 
+    def test__try_compile_no_another_exc_on_error(self):
```
Comment (iritkatriel) [inline Lib/test/test_dis.py]:
```suggestion
    def test__try_compile_no_context_exc_on_error(self):
```

File: Misc/NEWS.d/next/Library/2023-02-21-14-57-34.gh-issue-102114.uUDQzb.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Functions from :mod:`dis` module that accept source code string as argument now print shorter traceback if given string contains syntax/indentation error
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2023-02-21-14-57-34.gh-issue-102114.uUDQzb.rst]:
```suggestion
Functions in the :mod:`dis` module that accept a source code string as argument now
print a more concise traceback when the string contains a syntax or indentation error.
Patch by YOUR_NAME_HERE.
```

Comment (iritkatriel) [reply Misc/NEWS.d/next/Library/2023-02-21-14-57-34.gh-issue-102114.uUDQzb.rst]: That's just a suggestion. I can merge it as it is if you want.

Comment (iritkatriel) [reply Misc/NEWS.d/next/Library/2023-02-21-14-57-34.gh-issue-102114.uUDQzb.rst]: Don't forget to ``git pull`` because I merged main into this branch.

**Review by iritkatriel — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2023-02-21-14-57-34.gh-issue-102114.uUDQzb.rst (hunk 2 above)
Comment (chgnrdv) [reply Misc/NEWS.d/next/Library/2023-02-21-14-57-34.gh-issue-102114.uUDQzb.rst]: Thank you. I removed the signature as I find these changes pretty minor :)

File: Lib/dis.py [hunk 3]
```diff
@@ -66,6 +66,8 @@ def _try_compile(source, name):
     try:
         c = compile(source, name, 'eval')
     except SyntaxError:
+        c = None
+    if not c:
```
Comment (iritkatriel) [inline Lib/dis.py]: Actually one more issue - it's not obvious to me that compile can never return None (even if it's true now, how do we know it will continue to be true?)

Maybe better to have a boolean variable, set it to ``False`` before the ``try:`` and to ``True`` in the ``except:``, and then check it in the if?

Comment (iritkatriel) [reply Lib/dis.py]: Maybe I'm being silly, and it can't be None because it's supposed to be a code object or an ast. In that case maybe just change the check to "if c is None", instead of relying on the bool() value of the result.

Comment (iritkatriel) [inline Lib/dis.py]:
```suggestion
    if c is None:
```

Comment (chgnrdv) [reply Lib/dis.py]: For me it looks like call to `builtin_compile` always utilizes code object created by `_PyAssemble_MakeCodeObject` or AST returned by `PyAST_mod2obj` call, or sets exception and returns `NULL` if error is occurred, so its result can't be `None`.

**Review by iritkatriel — APPROVED** [review]
Yes, that's better.
